Changed kubelet cpu manager policy value to static

sureshmarikkannu opened the pull request:
If we want to allocate specific resources (Guaranteed QoS) for the POD's which requested guaranteed QoS (like UPF which needs CPU exclusively allocated for the bessd for better performance) then we should configure cpu manager policy as static. So if any POD's requires Guaranteed QoS then it can be handled, if not the best effort policy will take care of allocating resources for the POD's.

I came across an issue when i deploy multiple instances of UPF, the same CPU was allocated for both UPF pods which impacts the performance of the UPF significantly. So it would be good to make the default value as static.

gab-arrobo reviewed (approved):
@sureshmarikkannu, it would be good to point out that these changes (this patch) only have an effect when the resources in the `upf` are enabled. Thanks!

@llpeterson, please feel free to merge this PR and it works as expected. Thanks!

mbilal92 commented:
@gab-arrobo @sureshmarikkannu Should we add a flag to enable or disable this feature, or is it better to set it to static?
Also, would 1 core be sufficient for each `system-reserved` and `kube-reserved`?"

sureshmarikkannu commented:
> @gab-arrobo @sureshmarikkannu Should we add a flag to enable or disable this feature, or is it better to set it to static? Also, would 1 core be sufficient for each `system-reserved` and `kube-reserved`?"

@mbilal92 , 
1. The reason for assigning 1 cpu for these parameters is in a low core systems we should be able to deploy onramp without any issues. For example, in a 4 core system if we allocate 2 CPU's each for kube-reserved/system-reserved then we don't have other CPU's for POD's and other process. So i try to configure the bare minimum CPU which will work without any issues. So i tested with 1 CPU and didn't see any issues. i.e the reason i configured the value with 1 CPU each. Please find below the usage of these params,
"kube-reserved is meant to capture resource reservation for kubernetes system daemons like the kubelet, container runtime, node problem detector, etc. It is not meant to reserve resources for system daemons that are run as pods"

"system-reserved is meant to capture resource reservation for OS system daemons like sshd, udev, etc."

2. Regarding configurable option, yes it is possible. The reason we want to make this as default value is if the POD's not requesting Guaranteed QoS (not enabling Resource flags)  then this configuration will not impact those pods. The resources are allocated either Burstable OR BestEffort modes (same behavior as of today). This flag will impacts only the POD's which requests Guaranteed QoS.
  
3. If we want to make this as configurable option then we can add a parameter as below in vars/main.yml
k8s:
  rke2:
    version: v1.23.4+rke2r1
    config:
      **cpumanager:static (by default it will be configured as None)**
      token: purdue-k8s-rke2
      port: 9345
Then based on this flag we can copy the master files accordingly (we can have two file one with cpu policy static and other as normal like today we have). OR we can edit (by adding tasks in the rke install procedure) the existing files before copying it to the /etc/rancher folder.

mbilal92 commented:
@sureshmarikkannu Sounds good. If you have tested with 1 CPU and it works fine, let's keep it at 1.

Yes, let's use a flag for this, but let's call it `resources.enabled` (or something similar) that we can also use for Aether 5GC pods (UPF and others) in sd-core.yaml file 

```
  resources:
    enabled: false
```

sureshmarikkannu commented:
> @sureshmarikkannu Sounds good. If you have tested with 1 CPU and it works fine, let's keep it at 1.
> 
> Yes, let's use a flag for this, but let's call it `resources.enabled` (or something similar) that we can also use for Aether 5GC pods (UPF and others) in sd-core.yaml file
> 
> ```
>   resources:
>     enabled: false
> ```

@mbilal92 , I have added this option as configurable one as discussed. Please review PR https://github.com/opennetworkinglab/aether-onramp/pull/41 as well.

Please note that i have added cpu manager policy, kube cpu's and system cpu's as configurable parameters. So it can be tuned based on the available resources in the deployment system. The default option for cpu manager policy is 'none' which is the normal behaviour. If we need guaranteed QoS then we can enable it by configure it as 'static'. With this mechanism we can add other kubelet options in future based on need basis.

sureshmarikkannu commented:
> @sureshmarikkannu Sounds good. If you have tested with 1 CPU and it works fine, let's keep it at 1.
> 
> Yes, let's use a flag for this, but let's call it `resources.enabled` (or something similar) that we can also use for Aether 5GC pods (UPF and others) in sd-core.yaml file
> 
> ```
>   resources:
>     enabled: false
> ```

@mbilal92 , I have added this option as configurable one as discussed. Please review PR https://github.com/opennetworkinglab/aether-onramp/pull/41 as well.

Please note that i have added cpu manager policy, kube cpu's and system cpu's as configurable parameters. So it can be tuned based on the available resources in the deployment system. The default option for cpu manager policy is 'none' which is the normal behaviour. If we need guaranteed QoS then we can enable it by configure it as 'static'. With this mechanism we can add other kubelet options in future based on need basis.

sureshmarikkannu commented:
> @sureshmarikkannu Sounds good. If you have tested with 1 CPU and it works fine, let's keep it at 1.
> 
> Yes, let's use a flag for this, but let's call it `resources.enabled` (or something similar) that we can also use for Aether 5GC pods (UPF and others) in sd-core.yaml file
> 
> ```
>   resources:
>     enabled: false
> ```

@mbilal92 , I have added this option as configurable one as discussed. Please review PR https://github.com/opennetworkinglab/aether-onramp/pull/41 as well.

Please note that i have added cpu manager policy, kube cpu's and system cpu's as configurable parameters. So it can be tuned based on the available resources in the deployment system. The default option for cpu manager policy is 'none' which is the normal behaviour. If we need guaranteed QoS then we can enable it by configure it as 'static'. With this mechanism we can add other kubelet options in future based on need basis.

llpeterson commented:
Some of these changes run afoul of OnRamp's attempt to keep per-subsystem configuration separate from Aether-wide deployment parameters. Having an alternative `master-config.yaml` makes sense, but it should include all the required configuration changes, rather than "appending" additional changes in the playbook (as is done in this commit: https://github.com/opennetworkinglab/aether-k8s/pull/4/commits/8e9f746003bd78ef424d2592b18d96957b0c7972). If that means adding a second "config" file to the template directory, then that's ok; we just need to identify the config file to be used. (There is an Ansible variable to do this -- analogous to `core.values_file`, `amp.roc_models` and so on -- we just need to use it.) In general, there could be arbitrarily many modifications to `master-config.yaml` and we don't want to promote all of them to OnRamp-level status (or bury them in a playbook). They belong in the config file, which then gets named.

There is an OnRamp-level variable to consider -- `resources.enabled` -- but it needs to exist for both the k8s subsystem and the 5gc subsystem (which means it shows up in both the `k8s` and `core` blocks of `vars/main.yml`). We've recognized for a while that there are other "global" variables that users are currently forced to define multiple times; `data_iface` is a good example. We think it's time to add a `global` section to `vars/main.yml` that defines cross-subsystem variables. (Maybe we call it `aether` since it defines Aether-wide variables rather than per-subsystem variables.) Bilal is thinking this through to see if (a) there are any problems, and (b) there are other variables that should be global.
